lib/grandpa/message_handler.go

```diff
@@ -53,11 +53,20 @@ func (h *MessageHandler) handleMessage(from peer.ID, m GrandpaMessage) (network.
 	case *CommitMessage:
 		return nil, h.handleCommitMessage(msg)
 	case *NeighbourMessage:
+		// we can afford to not retry handling neighbour message, if it errors.
 		return nil, h.handleNeighbourMessage(msg)
 	case *CatchUpRequest:
 		return h.handleCatchUpRequest(msg)
 	case *CatchUpResponse:
-		return nil, h.handleCatchUpResponse(msg)
+		err := h.handleCatchUpResponse(msg)
+		if err == blocktree.ErrNodeNotFound {
+			// TODO: we are adding these messages to reprocess them again, but we
+			// haven't added code to reprocess them. Do that.
+			// Also, revisit if we need to add these message in synchronous manner
+			// or not. If not, change catchUpResponseMessages to a normal map.  #1531
```

Comment:
Why not have it as a standard `map` with a mutex instead? It adds some type safety compared to the `sync.Map`

Reply:
That is an option. I think I have used sync.Map, since it is tidier than using a mutex for every map. This logic could have multiple maps for each kind of message. But it is best to discuss this during the catchUpResponse related issue.

Also, you can always check the type of an object received sync.Map. I don't see if any of these approaches are clearly better than other, they are just different.

Reply:
Just create a custom type with a map and a mutex as fields. You can add typed get and put methods?

I don't think it's tidier to have no type safety.

There is a point in using `sync.Map` really only if we have a lot of reads on it and we can see it as a bottleneck (like `sync.Pool` really).

Reply:
addressed, using mutex now